drivers/led_strip/tlc5971.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright (c) 2021 Esco Medical ApS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_tlc5971
+
+#include <zephyr.h>
+#include <drivers/spi.h>
+#include <drivers/led_strip.h>
+#include <drivers/led_strip/tlc5971.h>
+#include <dt-bindings/led/led.h>
+#include <sys/util.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(tlc5971, CONFIG_LED_STRIP_LOG_LEVEL);
+
+struct tlc5971_config {
+	struct spi_dt_spec bus;
+	uint8_t num_pixels;
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+};
+
+struct tlc5971_data {
+	uint8_t gbc_color_1;
+	uint8_t gbc_color_2;
+	uint8_t gbc_color_3;
+	uint8_t control_data;
+	uint8_t *data_buffer;
+};
+
+/**
+ * @brief SPI operation word constant, SPI mode 0, CPOL = 0, CPHA = 0
+ */
+#define TLC5971_SPI_OPERATION (SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | SPI_WORD_SET(8))
+
+/**
+ * @brief Number of supported colors
+ */
+#define TLC5971_NUMBER_OF_COLORS 3
+
+/**
+ * @brief Number of RGB pixels per TLC5791 device
+ *
+ * The TLC5971 has 4x RGB outputs per device, where each RGB group constitues a pixel from this
+ * drivers point of view.
+ */
+#define TLC5971_PIXELS_PER_DEVICE 4
+
+/**
+ * @brief Length in bytes of data packet per TLC5791 device
+ */
+#define TLC5971_PACKET_LEN 28
+
+/**
+ * @brief write command for writing control data and GS data to internal registers
+ */
+#define TLC5971_WRITE_COMMAND (0x25)
+
+/**
+ * @brief GS reference clock select bit in FC data (0 = internal oscillator clock, 1 = SCKI clock).
+ */
+#define TLC5971_CTRL_BIT_EXTGCK (BIT(0))
+
+/**
+ * @brief GS reference clock edge select bit for OUTXn on-off timing control in FC data
+ */
+#define TLC5971_CTRL_BIT_OUTTMG (BIT(1))
+
+/**
+ * @brief Constant-current output enable bit in FC data (0 = output control enabled, 1 = blank).
+ */
+#define TLC5971_CTRL_BIT_BLANK (BIT(5))
+
+/**
+ * @brief Auto display repeat mode enable bit in FC data (0 = disabled, 1 = enabled).
+ */
+#define TLC5971_CTRL_BIT_DSPRPT (BIT(6))
+
+/**
+ * @brief Display timing reset mode enable bit in FC data (0 = disabled, 1 = enabled).
+ */
+#define TLC5971_CTRL_BIT_TMGRST (BIT(7))
+
+/**
+ * @brief Bit mask for control bits in data byte 27
+ */
+#define TLC5971_BYTE27_CTRL_MASK (TLC5971_CTRL_BIT_EXTGCK | TLC5971_CTRL_BIT_OUTTMG)
+
+/**
+ * @brief Bit mask for control bits in data byte 26
+ */
+#define TLC5971_BYTE26_CTRL_MASK                                                                   \
+	(TLC5971_CTRL_BIT_BLANK | TLC5971_CTRL_BIT_DSPRPT | TLC5971_CTRL_BIT_TMGRST)
+
+/**
+ * @brief create data byte 27 from control data
+ *
+ * @param control_data control bits
+ * @return uint8_t the serialized data byte 27
+ */
+static inline uint8_t tlc5971_data_byte27(uint8_t control_data)
+{
+	return (TLC5971_WRITE_COMMAND << 2) | (control_data & TLC5971_BYTE27_CTRL_MASK);
```

Comment:
Just a recommendation, try using `FIELD_PREP` macro for setting the bits paired with `GENMASK`.

Reply:
Done, although I'm not sure it really enhances readability. Let me know if I'm doing it wrong.

Reply:
For me, these macros help in clearly seeing the boundaries of the bit fields and they eliminate the need to provide both shift and masks values.

For instance in this function, now the user would know that the write command bits are situated between `7 to 2` of byte 27 and ctrl bits between `1 to 0` of byte 26.